g3/lowpan6_g3.c

```diff
@@ -618,126 +727,190 @@ lowpan6_hwaddr_to_addr(struct netif *netif, struct lowpan6_link_addr *addr)
 {
-  u8_t *puc, b;
-  s8_t i;
-  struct lowpan6_link_addr src, dest;
+  lowpan6_g3_data_t *ctx = (lowpan6_g3_data_t *) netif->state;
+  u8_t *puc, b, seq_num;
+  unsigned pos = 0;
+  int ret;
+  s16_t i;
   u16_t datagram_size = 0;
   u16_t datagram_offset, datagram_tag;
   struct lowpan6_reass_helper *lrh, *lrh_next, *lrh_prev = NULL;
+  struct lowpan6_g3_mesh_hdr mesh_header;
+  struct lowpan6_g3_routing_entry *entry;
+  struct lowpan6_link_addr next_hop, *originator = src;
 
   if (p == NULL) {
     return ERR_OK;
   }
 
   MIB2_STATS_NETIF_ADD(netif, ifinoctets, p->tot_len);
 
+  puc = (u8_t *)p->payload;
+  if (!lowpan6_g3_contains_lbp(puc) && indication->security_level < ctx->security_level) {
+    goto lowpan6_input_discard;
+  }
+
   if (p->len != p->tot_len) {
     /* for now, this needs a pbuf in one piece */
     goto lowpan6_input_discard;
   }
 
-  if (lowpan6_parse_iee802154_header(p, &src, &dest) != ERR_OK) {
-    goto lowpan6_input_discard;
+  /* Packet not for us. TODO: implement sniffer mode? */
+  if (!lowpan6_link_addr_cmp(dest, &ieee_802154_broadcast) &&
+      !lowpan6_link_addr_cmp(dest, &ctx->short_mac_addr) &&
+      !lowpan6_link_addr_cmp(dest, &ctx->extended_mac_addr)) {
+      goto lowpan6_input_discard;
   }
 
-  /* Check dispatch. */
-  puc = (u8_t *)p->payload;
+  b = *puc;
+  if ((b & 0xc0) == LOWPAN6_HEADER_MESH) {
+    if ((ret = lowpan6_parse_mesh_header(puc, p->len, &mesh_header)) < 0)
+      goto lowpan6_input_discard;
+
+    pos += ret;
+    originator = &mesh_header.originator;
+    /* G3 supports only mesh routing of short addresses */
+    if (mesh_header.originator.addr_len != 2 || mesh_header.final_dest.addr_len != 2)
+      goto lowpan6_input_discard;
+
+    if (mesh_header.hops_left == 0)
+      goto lowpan6_input_discard;
+
+    if (lowpan6_link_addr_cmp(dest, &ieee_802154_broadcast)) {
+      /* Multicast case */
+      if ((ret = lowpan6_g3_parse_bc0_header(puc + pos, p->len - pos, &seq_num)) < 0)
+        goto lowpan6_input_discard;
 
+      pos += ret;
+      /* Neither broadcast frame nor destined to a group, to which we belong */
+      if (lowpan6_link_addr_to_u16(&mesh_header.final_dest) != PP_HTONS(LOWPAN6_BROADCAST_ADDR) &&
+          lowpan6_g3_group_table_lookup(lowpan6_link_addr_to_u16(&mesh_header.final_dest)) == NULL) {
+          goto lowpan6_input_discard;
+      }
+
+      /* Have we seen this frame? */
+      if (lowpan6_g3_broadcast_log_table_lookup(lowpan6_link_addr_to_u16(&mesh_header.originator), seq_num) != NULL ||
+          memcmp(mesh_header.originator.addr, ctx->short_mac_addr.addr, 2) == 0) {
+          goto lowpan6_input_discard;
+      }
+
+      /* Accept frame */
+      lowpan6_g3_broadcast_log_table_add(lowpan6_link_addr_to_u16(&mesh_header.originator), seq_num);
+
+      /* Check if we shall retransmit this frame */
+      if (mesh_header.hops_left > 1) {
```

Comment:
Maybe I have done it a little bit too implicit, but the resulting behavior conforms to the specs. The algorithm described in the specs says to decrement to `hops_left` field and then check: `hops_left > 0`. 
I guess, I did it this way in order to avoid decrementing `hops_left` two times, as the `mesh_header` does not directly
point to the data buffer. I parse the Mesh Header and save it in a non-packed structure instead of simply casting the pointer to a packed structure, because the RFC allows different formats of the mesh header. But G3 allows only one format with short 16-bites addresses. Maybe I should stick to the G3 definition.

Reply:
ok, maybe some description in comment could be helpful (that it's pre-decrement value) because otherwise it looks suspicious (like an error ;)).

Not a merge blocker.